File: platform/WinForms/SweetEditor/EditorSettings.cs
namespace SweetEditor {
	/// <summary>
	/// Centralized configuration for <see cref="EditorControl"/>.
	/// <para>
	/// Obtain via <see cref="EditorControl.Settings"/>. All setters take effect immediately.
	/// </para>
	/// </summary>
	public class EditorSettings {

		private readonly EditorControl editor;

		private float scale = 1.0f;
		private FoldArrowMode foldArrowMode = FoldArrowMode.ALWAYS;
		private WrapMode wrapMode = WrapMode.NONE;
		private float lineSpacingAdd = 0f;
		private float lineSpacingMult = 1.0f;
		private float contentStartPadding = 0f;
		private bool showSplitLine = true;
		private CurrentLineRenderMode currentLineRenderMode = CurrentLineRenderMode.BACKGROUND;
		private AutoIndentMode autoIndentMode = AutoIndentMode.NONE;
		private bool readOnly = false;
		private int maxGutterIcons = 0;
		private int decorationScrollRefreshMinIntervalMs = 16;
		private float decorationOverscanViewportMultiplier = 1.5f;

		internal EditorSettings(EditorControl editor) {
			this.editor = editor;
		}

		/// <summary>Sets editor scale.</summary>
		public void SetScale(float scale) {
			this.scale = scale;
			editor.EditorCoreInternal.SetScale(scale);
			editor.SyncPlatformScaleInternal(scale);
			editor.Flush();
		}

		/// <summary>Gets editor scale.</summary>
		public float GetScale() => scale;

		/// <summary>Sets fold arrow mode.</summary>
		public void SetFoldArrowMode(FoldArrowMode mode) {
			foldArrowMode = mode;
			editor.EditorCoreInternal.SetFoldArrowMode((int)mode);
		}

		/// <summary>Gets fold arrow mode.</summary>
		public FoldArrowMode GetFoldArrowMode() => foldArrowMode;

		/// <summary>Sets wrap mode.</summary>
		public void SetWrapMode(WrapMode mode) {
			wrapMode = mode;
			editor.EditorCoreInternal.SetWrapMode((int)mode);
			editor.Flush();
		}

		/// <summary>Gets wrap mode.</summary>
		public WrapMode GetWrapMode() => wrapMode;

		/// <summary>Sets line spacing.</summary>
		public void SetLineSpacing(float add, float mult) {
			lineSpacingAdd = add;
			lineSpacingMult = mult;
			editor.EditorCoreInternal.SetLineSpacing(add, mult);
			editor.Flush();
		}

		/// <summary>Gets line spacing add.</summary>
		public float GetLineSpacingAdd() => lineSpacingAdd;

		/// <summary>Gets line spacing multiplier.</summary>
		public float GetLineSpacingMult() => lineSpacingMult;

		/// <summary>Sets extra horizontal padding between gutter split and text content start.</summary>
		public void SetContentStartPadding(float padding) {
			contentStartPadding = System.Math.Max(0f, padding);
			editor.EditorCoreInternal.SetContentStartPadding(contentStartPadding);
			editor.Flush();
		}

		/// <summary>Gets content start padding.</summary>
		public float GetContentStartPadding() => contentStartPadding;

		/// <summary>Sets whether gutter split line should be rendered.</summary>
		public void SetShowSplitLine(bool show) {
			showSplitLine = show;
			editor.EditorCoreInternal.SetShowSplitLine(show);
			editor.Flush();
		}

		/// <summary>Gets whether gutter split line should be rendered.</summary>
		public bool IsShowSplitLine() => showSplitLine;

		/// <summary>Sets current line render mode.</summary>
		public void SetCurrentLineRenderMode(CurrentLineRenderMode mode) {
			currentLineRenderMode = mode;
			editor.EditorCoreInternal.SetCurrentLineRenderMode(mode);
			editor.Flush();
		}

		/// <summary>Gets current line render mode.</summary>
		public CurrentLineRenderMode GetCurrentLineRenderMode() => currentLineRenderMode;

		/// <summary>Sets auto indent mode.</summary>
		public void SetAutoIndentMode(AutoIndentMode mode) {
			autoIndentMode = mode;
			editor.EditorCoreInternal.SetAutoIndentMode((int)mode);
		}

		/// <summary>Gets auto indent mode.</summary>
		public AutoIndentMode GetAutoIndentMode() => autoIndentMode;

		/// <summary>Sets read-only mode.</summary>
		public void SetReadOnly(bool readOnly) {
			this.readOnly = readOnly;
			editor.EditorCoreInternal.SetReadOnly(readOnly);
		}

		/// <summary>Gets read-only mode.</summary>
		public bool IsReadOnly() => readOnly;

		/// <summary>Sets max gutter icons.</summary>
		public void SetMaxGutterIcons(int count) {
			maxGutterIcons = count;
			editor.EditorCoreInternal.SetMaxGutterIcons(count);
		}

		/// <summary>Gets max gutter icons.</summary>
		public int GetMaxGutterIcons() => maxGutterIcons;

		/// <summary>Sets minimum interval for scroll-triggered decoration refresh (milliseconds).</summary>
		public void SetDecorationScrollRefreshMinIntervalMs(int intervalMs) {
			decorationScrollRefreshMinIntervalMs = System.Math.Max(0, intervalMs);
			editor.RequestDecorationRefresh();
		}

		/// <summary>Gets minimum interval for scroll-triggered decoration refresh (milliseconds).</summary>
		public int GetDecorationScrollRefreshMinIntervalMs() => decorationScrollRefreshMinIntervalMs;

		/// <summary>Sets decoration overscan multiplier relative to viewport line count.</summary>
		public void SetDecorationOverscanViewportMultiplier(float multiplier) {
			decorationOverscanViewportMultiplier = System.Math.Max(0f, multiplier);
			editor.RequestDecorationRefresh();
		}

		/// <summary>Gets decoration overscan multiplier relative to viewport line count.</summary>
		public float GetDecorationOverscanViewportMultiplier() => decorationOverscanViewportMultiplier;
	}
}

File: platform/Swing/sweeteditor/src/main/java/com/qiplat/sweeteditor/EditorSettings.java
package com.qiplat.sweeteditor;

import com.qiplat.sweeteditor.core.foundation.AutoIndentMode;
import com.qiplat.sweeteditor.core.foundation.CurrentLineRenderMode;
import com.qiplat.sweeteditor.core.foundation.FoldArrowMode;
import com.qiplat.sweeteditor.core.foundation.WrapMode;

/**
 * Centralized configuration for {@link SweetEditor}.
 * <p>
 * Obtain via {@link SweetEditor#getSettings()}. All setters take effect immediately.
 */
public class EditorSettings {

    private final SweetEditor mEditor;

    private float mScale = 1.0f;
    private FoldArrowMode mFoldArrowMode = FoldArrowMode.ALWAYS;
    private WrapMode mWrapMode = WrapMode.NONE;
    private float mLineSpacingAdd = 0f;
    private float mLineSpacingMult = 1.0f;
    private float mContentStartPadding = 0f;
    private boolean mShowSplitLine = true;
    private boolean mGutterSticky = true;
    private CurrentLineRenderMode mCurrentLineRenderMode = CurrentLineRenderMode.BACKGROUND;
    private AutoIndentMode mAutoIndentMode = AutoIndentMode.NONE;
    private boolean mReadOnly = false;
    private int mMaxGutterIcons = 0;
    private int mDecorationScrollRefreshMinIntervalMs = 16;
    private float mDecorationOverscanViewportMultiplier = 1.5f;

    EditorSettings(SweetEditor editor) {
        mEditor = editor;
    }

    public void setScale(float scale) {
        mScale = scale;
        mEditor.getEditorCore().setScale(scale);
        mEditor.syncPlatformScale(scale);
        mEditor.flush();
    }

    public float getScale() {
        return mScale;
    }

    public void setFoldArrowMode(FoldArrowMode mode) {
        mFoldArrowMode = mode;
        mEditor.getEditorCore().setFoldArrowMode(mode.value);
    }

    public FoldArrowMode getFoldArrowMode() {
        return mFoldArrowMode;
    }

    public void setWrapMode(WrapMode mode) {
        mWrapMode = mode;
        mEditor.getEditorCore().setWrapMode(mode.value);
        mEditor.flush();
    }

    public WrapMode getWrapMode() {
        return mWrapMode;
    }

    public void setLineSpacing(float add, float mult) {
        mLineSpacingAdd = add;
        mLineSpacingMult = mult;
        mEditor.getEditorCore().setLineSpacing(add, mult);
        mEditor.flush();
    }

    public float getLineSpacingAdd() {
        return mLineSpacingAdd;
    }

    public float getLineSpacingMult() {
        return mLineSpacingMult;
    }

    public void setContentStartPadding(float padding) {
        mContentStartPadding = Math.max(0f, padding);
        mEditor.getEditorCore().setContentStartPadding(mContentStartPadding);
        mEditor.flush();
    }

    public float getContentStartPadding() {
        return mContentStartPadding;
    }

    public void setShowSplitLine(boolean show) {
        mShowSplitLine = show;
        mEditor.getEditorCore().setShowSplitLine(show);
        mEditor.flush();
    }

    public boolean isShowSplitLine() {
        return mShowSplitLine;
    }

    public void setGutterSticky(boolean sticky) {
        mGutterSticky = sticky;
        mEditor.getEditorCore().setGutterSticky(sticky);
        mEditor.flush();
    }

    public boolean isGutterSticky() {
        return mGutterSticky;
    }

    public void setCurrentLineRenderMode(CurrentLineRenderMode mode) {
        mCurrentLineRenderMode = mode;
        mEditor.getEditorCore().setCurrentLineRenderMode(mode.value);
        mEditor.flush();
    }

    public CurrentLineRenderMode getCurrentLineRenderMode() {
        return mCurrentLineRenderMode;
    }

    public void setAutoIndentMode(AutoIndentMode mode) {
        mAutoIndentMode = mode;
        mEditor.getEditorCore().setAutoIndentMode(mode.value);
    }

    public AutoIndentMode getAutoIndentMode() {
        return mAutoIndentMode;
    }

    public void setReadOnly(boolean readOnly) {
        mReadOnly = readOnly;
        mEditor.getEditorCore().setReadOnly(readOnly);
    }

    public boolean isReadOnly() {
        return mReadOnly;
    }

    public void setMaxGutterIcons(int count) {
        mMaxGutterIcons = count;
        mEditor.getEditorCore().setMaxGutterIcons(count);
    }

    public int getMaxGutterIcons() {
        return mMaxGutterIcons;
    }

    public void setDecorationScrollRefreshMinIntervalMs(int intervalMs) {
        mDecorationScrollRefreshMinIntervalMs = Math.max(0, intervalMs);
        mEditor.requestDecorationRefresh();
    }

    public int getDecorationScrollRefreshMinIntervalMs() {
        return mDecorationScrollRefreshMinIntervalMs;
    }

    public void setDecorationOverscanViewportMultiplier(float multiplier) {
        mDecorationOverscanViewportMultiplier = Math.max(0f, multiplier);
        mEditor.requestDecorationRefresh();
    }

    public float getDecorationOverscanViewportMultiplier() {
        return mDecorationOverscanViewportMultiplier;
    }
}
